spellcheck.lua

```diff
@@ -3,32 +3,106 @@
-elseif os.execute("type aspell "..supress_output) then
+elseif os.execute("type aspell"..supress_output) then
 	spellcheck.cmd = "aspell -l %s -a"
 	spellcheck.list_cmd = "aspell list -l %s -a"
-elseif os.execute("type hunspell "..supress_output) then
+elseif os.execute("type hunspell"..supress_output) then
 	spellcheck.cmd = "hunspell -d %s"
 	spellcheck.list_cmd = "hunspell -l -d %s"
 else
    return nil
 end
 
 spellcheck.typo_style = "fore:red"
-spellcheck.enabled = {}
+spellcheck.check_full_viewport = {}
+spellcheck.disable_syntax_awareness = false
 
+spellcheck.check_tokens = {
+	[vis.lexers.STRING] = true,
+	[vis.lexers.COMMENT] = true
+}
+
+
+-- Return nil or a string of misspelled word in a specific file range or text
+-- by calling the spellchecker's list command.
+-- If given a range we will use vis:pipe to get our typos from the spellchecker.
+-- If a string was passed we call the spellchecker ourself and redirect its stdout
+-- to a temporary file. See http://lua-users.org/lists/lua-l/2007-10/msg00189.html.
+-- The returned string consists of each misspell followed by a newline.
+local function get_typos(range_or_text)
+	local cmd = spellcheck.list_cmd:format(spellcheck.lang)
+	local typos = nil
+	if type(range_or_text) == "string" then
+		local text = range_or_text
+		local tmp_name = os.tmpname()
+		local full_cmd = cmd .. "> " .. tmp_name .. supress_stderr
+		local proc = assert(io.popen(full_cmd, "w"))
+		proc:write(text)
+		-- this error detection may need lua5.2
+		local success, reason, exit_code = proc:close()
+		if not success then
+			vis:info("calling " .. cmd .. " failed ("..exit_code..")")
+			return nil
+		end
+
+		local tmp_file = assert(io.open(tmp_name, "r"))
+		typos = tmp_file:read("*a")
+		tmp_file:close()
+		os.remove(tmp_name)
+	else
+		local range = range_or_text
+		local ret, so, se = vis:pipe(vis.win.file, range, cmd)
+
+		if ret ~= 0 then
+			vis:info("calling " .. cmd .. " failed ("..ret..")")
+			return nil
+		end
+		typos = so
+	end
+
+	return typos
+end
+
+-- plugin global list of ignored typos
 local ignored = {}
 
+-- Return an iterator over all not ignored typos and their positions in text.
+-- The returned iterator is a self contained statefull iterator function closure.
+-- Which will return the next typo and its start and finish in the text, starting by 1.
+local function typo_iter(text, typos, ignored)
+	local index = 0
+	local unfiltered_iterator, iter_state = typos:gmatch("(.-)\n")
+	return function(foo, bar)
+		repeat
+			typo = unfiltered_iterator(iter_state)
+		until(not typo or not ignored[typo])
+
+		if typo then
+			-- to prevent typos from being found in correct words before them
+			-- ("stuff stuf", "broken ok", ...)
+			-- we match typos only when they are enclosed in non-letter characters.
+			local start, finish = text:find("[%A]" .. typo .. "[%A]", index)
+			index = finish
+
+			-- ignore the first and last non letter character
+			return typo, start + 1, finish - 1
```

Comment:
Error: start and finish can be nil value which stops the preforming of arithmetic. Should maybe be a if before to check start and finish values.